controllers/etcdbackup_controller_new.go

```diff
@@ -0,0 +1,267 @@
+	case state.backup.Status.Phase == etcdv1alpha1.EtcdBackupPhaseFailed:
+		log.Info("Backup has failed. Ignoring.")
+	case state.backup.Status.Phase == etcdv1alpha1.EtcdBackupPhaseCompleted:
+		log.Info("Backup has completed. Ignoring.")
+	case state.actual.serviceAccount == nil:
+		log.Info("Service account does not exist. Creating.")
+		action = &CreateRuntimeObject{client: r.Client, obj: state.desired.serviceAccount}
+	case state.actual.pod == nil:
+		log.Info("Pod does not exist. Creating.")
+		action = &CreateRuntimeObject{client: r.Client, obj: state.desired.pod}
+	case state.actual.pod.Status.Phase == corev1.PodFailed:
+		log.Info("Backup agent failed. Updating status.")
+		new := state.backup.DeepCopy()
+		new.Status.Phase = etcdv1alpha1.EtcdBackupPhaseFailed
+		action = &PatchStatus{client: r.Client, original: state.backup, new: new}
+		event = &reconcilerevent.BackupFailed{For: state.backup}
+	case state.actual.pod.Status.Phase == corev1.PodSucceeded:
+		log.Info("Backup agent succeeded. Updating status.")
+		new := state.backup.DeepCopy()
+		new.Status.Phase = etcdv1alpha1.EtcdBackupPhaseCompleted
+		action = &PatchStatus{client: r.Client, original: state.backup, new: new}
+		event = &reconcilerevent.BackupSucceeded{For: state.backup}
+	}
+
+	// Execute the action
+	if action != nil {
+		err := action.Execute(ctx)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error executing action: %s", err)
+		}
+		// Generate a generic event for CreateRuntimeObject actions
+		if event == nil {
+			switch o := action.(type) {
+			case *CreateRuntimeObject:
+				event = &reconcilerevent.ObjectCreatedEvent{Log: log, For: state.backup, Object: o.obj}
+			}
+		}
+	}
+
+	// Record an event
+	if event != nil {
+		event.Record(r.Recorder)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EtcdBackupReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&etcdv1alpha1.EtcdBackup{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&corev1.Pod{}).
+		Complete(r)
+}
+
+// serviceAccountForBackup creates a restriced service-account by the backup-agent pod.
+func serviceAccountForBackup(backup *etcdv1alpha1.EtcdBackup) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backup.Name,
+			Namespace: backup.Namespace,
+		},
+	}
+}
+
+// podForBackup creates a pod for running the backup-agent image.
+// It does not need to interact with the API and should not have permissions to
+// do so.
+func podForBackup(backup *etcdv1alpha1.EtcdBackup, image, proxyURL string) (*corev1.Pod, error) {
+	tmpl, err := template.New("template").Parse(backup.Spec.Destination.ObjectURL)
+	if err != nil {
+		return nil, fmt.Errorf("error %q parsing object URL template", err)
+	}
+	var objectURL strings.Builder
+	if err := tmpl.Execute(&objectURL, backup); err != nil {
+		return nil, fmt.Errorf("error %q executing template", err)
+	}
```

Comment:
I've just added this (hack?) to allow scheduled backups. It allows template variables to be included in the objectURL, to ensure unique backup names.

Reply:
Discussed in person.
I've changed the field name to `objectURLTemplate` and beefed up the documentation.
I'm happy to alter the API if users find it inadequate.